dartagnan/src/main/java/com/dat3m/dartagnan/encoding/ExpressionEncoder.java

```diff
@@ -0,0 +1,444 @@
+        return booleanFormulaManager.makeVariable(Integer.toString(bNonDet.hashCode()));
+    }
+
+    @Override
+    public Formula visit(IValue iValue) {
+        return iValue.getPrecision() > 0
+                ? bitvectorFormulaManager.makeBitvector(iValue.getPrecision(), iValue.getValue())
+                : integerFormulaManager.makeNumber(iValue.getValue());
+    }
+
+    @Override
+    public Formula visit(IExprBin iBin) {
+        Formula lhs = encodeInteger(iBin.getLHS());
+        Formula rhs = encodeInteger(iBin.getRHS());
+        // Some SMT solvers do not support certain theories.
+        // Calling the constructor of the manager in such solvers results in an Exception.
+        // Thus we initialize the manager inside the branches
+        if (lhs instanceof NumeralFormula.IntegerFormula && rhs instanceof NumeralFormula.IntegerFormula) {
+            NumeralFormula.IntegerFormula i1 = (NumeralFormula.IntegerFormula) lhs;
+            NumeralFormula.IntegerFormula i2 = (NumeralFormula.IntegerFormula) rhs;
+            switch (iBin.getOp()) {
+                case PLUS:
+                    return integerFormulaManager.add(i1, i2);
+                case MINUS:
+                    return integerFormulaManager.subtract(i1, i2);
+                case MULT:
+                    return integerFormulaManager.multiply(i1, i2);
+                case DIV:
+                case UDIV:
+                    return integerFormulaManager.divide(i1, i2);
+                case MOD:
+                    return integerFormulaManager.modulo(i1, i2);
+                case AND:
+                    return bitvectorFormulaManager.toIntegerFormula(bitvectorFormulaManager.and(bitvectorFormulaManager.makeBitvector(32, i1), bitvectorFormulaManager.makeBitvector(32, i2)), false);
+                case OR:
+                    return bitvectorFormulaManager.toIntegerFormula(bitvectorFormulaManager.or(bitvectorFormulaManager.makeBitvector(32, i1), bitvectorFormulaManager.makeBitvector(32, i2)), false);
+                case XOR:
+                    return bitvectorFormulaManager.toIntegerFormula(bitvectorFormulaManager.xor(bitvectorFormulaManager.makeBitvector(32, i1), bitvectorFormulaManager.makeBitvector(32, i2)), false);
+                case L_SHIFT:
+                    return bitvectorFormulaManager.toIntegerFormula(bitvectorFormulaManager.shiftLeft(bitvectorFormulaManager.makeBitvector(32, i1), bitvectorFormulaManager.makeBitvector(32, i2)), false);
+                case R_SHIFT:
+                    return bitvectorFormulaManager.toIntegerFormula(bitvectorFormulaManager.shiftRight(bitvectorFormulaManager.makeBitvector(32, i1), bitvectorFormulaManager.makeBitvector(32, i2), false), false);
+                case AR_SHIFT:
+                    return bitvectorFormulaManager.toIntegerFormula(bitvectorFormulaManager.shiftRight(bitvectorFormulaManager.makeBitvector(32, i1), bitvectorFormulaManager.makeBitvector(32, i2), true), false);
```

Comment:
Why 32 and not arch-precision?

Reply:
This seems to be an older issue, since I copied it from `IOpBin`.  It requires the arch precision to be positive / bounded.

Reply:
Arch precision cannot be used here, because this code is only executed when `arch-precision = -1` holds.
More precisely, the code runs when we do an integer-based encoding but run into BV operations.
In that case, we need to convert the integers to BV, but have no idea what bit-width we should use (since arch-precision=-1), though I think it woukd be more safe to default to 64 bits rather than 32.

On the other hand, we can keep it like this for now as the issue will go away with the coming rework.